demoapps/passwordlogin/java/org/openyolo/demoapp/passwordlogin/userdata/UserRepository.java

```diff
@@ -0,0 +1,118 @@
+ * limitations under the License.
+ */
+
+package org.openyolo.demoapp.passwordlogin.userdata;
+
+import android.arch.persistence.room.Room;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.content.SharedPreferences.Editor;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.WorkerThread;
+import org.openyolo.demoapp.passwordlogin.OpenYoloDemoApplication;
+
+/**
+ * A local implementation of UserDataSource for the purposes of this demo. This could be
+ * replaced with a more realistic implementation that makes network requests in order to
+ * authenticate the user.
+ */
+@WorkerThread
+public class UserRepository implements UserDataSource {
+
+    private static final String KEY_CURRENT_USER = "user";
+
+    private final UserDatabase mUserDatabase;
+    private final SharedPreferences mSharedPrefs;
+
+    /**
+     * Creates the user repository, with the required application reference.
+     */
+    public UserRepository(@NonNull OpenYoloDemoApplication application) {
+        mUserDatabase = Room.databaseBuilder(application, UserDatabase.class, "userdb")
+                .build();
+        mSharedPrefs = application.getSharedPreferences("authInfo", Context.MODE_PRIVATE);
+    }
+
+    @Override
+    @Nullable
+    public User getCurrentUser() {
+        String currentUserEmail = getCurrentUserEmail();
+        if (currentUserEmail == null) {
+            return null;
+        }
+
+        return mUserDatabase.userDao().getUserByEmail(currentUserEmail);
+    }
+
+    @Override
+    public boolean isExistingAccount(String email) {
+        return mUserDatabase.userDao().getUserByEmail(email) != null;
+    }
+
+    @Override
+    public boolean createPasswordAccount(
+            @NonNull String email,
+            String name,
+            String profilePictureUri,
+            String password) {
+        if (isExistingAccount(email)) {
+            return false;
+        }
+
+        String salt = HashUtil.generateSalt();
+        String hash = HashUtil.hashPassword(salt, password);
+
+        mUserDatabase.userDao().createUser(new User(email, name, salt, hash, profilePictureUri));
+        setCurrentUserEmail(email);
+        return true;
+    }
+
+    @Override
+    public boolean authWithPassword(String email, String password) {
+        User user = mUserDatabase.userDao().getUserByEmail(email);
+        if (user == null) {
+            return false;
+        }
+
+        String salt = user.getPasswordSalt();
+        String hash = HashUtil.hashPassword(salt, password);
+
+        if (hash.equals(user.getPasswordHash())) {
+            setCurrentUserEmail(email);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void signOut() {
+        setCurrentUserEmail(null);
+    }
+
+    private String getCurrentUserEmail() {
```

Comment:
To stay consistent add @NonNull annotation.

Reply:
Done.